src/dune_engine/build_system.ml

```diff
@@ -2272,25 +2272,40 @@ let prefix_rules (prefix : unit Action_builder.t) ~f =
   in
   res
 
-let report_early_exn ~report_error exn =
+let caused_by_cancellation (exn : Exn_with_backtrace.t) =
+  match exn.exn with
+  | Scheduler.Run.Build_cancelled -> true
+  | Memo.Error.E err -> (
+    match Memo.Error.get err with
```

Comment:
To avoid dealing with the optional wrapping, we could change `Memo` as follow:

- make `handle_error_no_raise` take a `Memo.Error.t` and a backtrace rather than an `Exn_with_backtrace.t`
- make `Memo.Build.run*` only raise `Memo.Error.E`

Reply:
I think it is already the case that the top-level `Memo.Build.run*` only ever raise `Memo.Error.E`. We could probably make them return it instead of raising it, even.

For the first bullet-point, agreed. I was even thinking of including the backtrace into `Memo.Error.t`, but that's a separate idea.

(I'm assuming you're saying these as ideas for the future PRs, so I'm merging this one)